internal/demo/script.go

```diff
@@ -3,28 +3,183 @@ package demo
 import (
 	"context"
 	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
 
+	"github.com/pkg/errors"
+	"github.com/windmilleng/tilt/internal/engine"
 	"github.com/windmilleng/tilt/internal/hud"
+	"github.com/windmilleng/tilt/internal/hud/client"
 	"github.com/windmilleng/tilt/internal/k8s"
-	"github.com/windmilleng/tilt/internal/model"
+	"github.com/windmilleng/tilt/internal/logger"
+	"github.com/windmilleng/tilt/internal/output"
+	"github.com/windmilleng/tilt/internal/store"
+	"github.com/windmilleng/tilt/internal/tiltfile"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/api/core/v1"
 )
 
 // Runs the demo script
 type Script struct {
 	hud   hud.HeadsUpDisplay
-	upper model.ManifestCreator
+	upper engine.Upper
+	store *store.Store
 	env   k8s.Env
+
+	readTiltfileCh chan string
+	podMonitor     *podMonitor
+}
+
+func NewScript(upper engine.Upper, hud hud.HeadsUpDisplay, env k8s.Env, st *store.Store) Script {
+	s := Script{
+		upper:          upper,
+		hud:            hud,
+		env:            env,
+		readTiltfileCh: make(chan string),
+		podMonitor:     &podMonitor{podsReadyCh: make(chan bool)},
+		store:          st,
+	}
+	st.AddSubscriber(s.podMonitor)
+	return s
+}
+
+type podMonitor struct {
+	podsReady   bool
+	podsReadyCh chan bool
+}
+
+func (m *podMonitor) OnChange(ctx context.Context, store *store.Store) {
+	state := store.RLockState()
+	hasPods := false
+	allPodsReady := true
+	for _, ms := range state.ManifestStates {
+		if ms.Pod.PodID != "" {
+			hasPods = true
+		}
+
+		if ms.Pod.Phase != v1.PodRunning {
+			allPodsReady = false
+		}
+	}
+
+	defer store.RUnlockState()
+
+	podsReady := hasPods && allPodsReady
+	if podsReady != m.podsReady {
+		m.podsReady = podsReady
+		m.podsReadyCh <- podsReady
+	}
 }
 
-func NewScript(upper model.ManifestCreator, hud hud.HeadsUpDisplay, env k8s.Env) Script {
-	return Script{
-		upper: upper,
-		hud:   hud,
-		env:   env,
+func (m *podMonitor) waitUntilPodsReady(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case ready := <-m.podsReadyCh:
+			if ready {
+				return nil
+			}
+		}
 	}
 }
 
 func (s Script) Run(ctx context.Context) error {
-	fmt.Println("TODO(nick): implement demo")
+	if !s.env.IsLocalCluster() {
+		_, _ = fmt.Fprintf(os.Stderr, "tilt demo mode only supports docker-for-desktop or Minikube\n")
```

Comment:
Maybe some additional verbiage here for docker-for-desktop? It seems less obvious and google-eable than Minikube

Reply:
good point! i changed it to Docker For Mac